src/main/java/org/hisp/dhis/rules/parser/expression/function/OperatorLogicalAnd.java

```diff
@@ -0,0 +1,75 @@
+package org.hisp.dhis.rules.parser.expression.function;
+
+/*
+ * Copyright (c) 2004-2020, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import org.hisp.dhis.antlr.AntlrExpressionVisitor;
+import org.hisp.dhis.antlr.operator.AntlrOperatorLogicalAnd;
+import org.hisp.dhis.parser.expression.antlr.ExpressionParser;
+import org.hisp.dhis.rules.parser.expression.CommonExpressionVisitor;
+
+/**
+ * Logical operator: And
+ * <pre>
+ *
+ * Truth table (same as for SQL):
+ *
+ *       A      B    A and B
+ *     -----  -----  -------
+ *     null   null    null
+ *     null   false   null
+ *     null   true    null
+ *
+ *     false  null    false
+ *     false  false   false
+ *     false  true    false
+ *
+ *     true   null    null
+ *     true   false   false
+ *     true   true    true
+ * </pre>
+ *
+ * @author Jim Grace
+ */
+public class OperatorLogicalAnd extends ScalarFunctionToEvaluate
+{
+    @Override
+    public Object evaluate( ExpressionParser.ExprContext ctx, CommonExpressionVisitor visitor )
+    {
+        Boolean value0 = visitor.castBooleanVisit( ctx.expr( 0 ) );
+        Boolean value1 = visitor.castBooleanVisit( ctx.expr( 1 ) );
+
+        return value0 != null && value0 ? value1 : true;
+    }
+
+    @Override
+    public Object getDescription( ExpressionParser.ExprContext ctx, CommonExpressionVisitor visitor )
+    {
+        return evaluate( ctx, (AntlrExpressionVisitor) visitor );
+    }
```

Comment:
You are right that when making a description, you need to evaluate both sides of the "and" operator.

The way you have the evaluate() method now, it always evaluates the right side even if the left side is false. This changes the current behaviour in AntlrOperatorLogicalAnd, where the second expr is evaluated only when the first expr is true. Does this make a difference for program rules? It does make a difference for many other types of expressions.

Another way you could do this is to have evaluate() just call AntlrOperatorLogicalAnd().evaluate(), similar to what you have done in other classes. Then the getDescription() method could be something like:

        visitor.castBooleanVisit( ctx.expr( 0 ) );
        visitor.castBooleanVisit( ctx.expr( 1 ) );

        return CommonExpressionVisitor.DEFAULT_BOOLEAN_VALUE;

This would keep the same evaluate() functionality. It also means that the truth table implementation is done only in one place (in AntlrOperatorLogicalAnd), so the comments about the truth table could be removed from this class.

Reply:
resolved

Reply:
For consistency with your other operators (and for good practice -- even though it's not done in many parts of DHIS2), I suggest you add back in the one JavaDoc line:
 * Logical operator: And
Likewise with Or and Not.